File: core/src/main/java/com/roboxue/niffler/javaDSL/Niffler.java
package com.roboxue.niffler.javaDSL;

import com.google.common.collect.Iterables;
import com.roboxue.niffler.DataFlow;
import com.roboxue.niffler.ExecutionStateTracker;
import com.roboxue.niffler.LanguageBridge;
import com.roboxue.niffler.Logic;
import com.roboxue.niffler.Token;
import com.roboxue.niffler.execution.AsyncExecution;
import com.roboxue.niffler.execution.ExecutionLogger;

import javax.annotation.Nullable;
import java.util.logging.Level;
import java.util.logging.Logger;

public interface Niffler {

    Iterable<DataFlow<?>> getDataFlows();

    default <T> AsyncExecution<T> asyncRun(Token<T> token, Iterable<DataFlow<?>> extraFlow,
            ExecutionStateTracker sc, @Nullable ExecutionLogger logger) {
        return LanguageBridge.asyncRun(this, token, extraFlow, sc, logger);
    }

    default void printGraph(Logger logger, boolean useCodeName) {
        new Logic(LanguageBridge.javaIterableAsSeq(getDataFlows()))
                .printFlowChart(logger, Level.INFO, useCodeName);
    }

    default void printGraph(Logger logger, boolean useCodeName, Iterable<DataFlow<?>> extraFlows) {
        new Logic(LanguageBridge.javaIterableAsSeq(Iterables.concat(getDataFlows(), extraFlows)))
                .printFlowChart(logger, Level.INFO, useCodeName);
    }
}

File: core/src/main/scala/com/roboxue/niffler/scalaDSL/Niffler.scala
package com.roboxue.niffler.scalaDSL

import java.util.logging.{Level, Logger}

import com.roboxue.niffler._
import com.roboxue.niffler.execution.{AsyncExecution, ExecutionLogger}

/**
  * @author robert.xue
  * @since 7/15/18
  */
object Niffler {}

trait Niffler {
  self =>
  def dataFlows: Iterable[DataFlow[_]]

  def asyncRun[T](token: Token[T], extraFlow: Iterable[DataFlow[_]] = Iterable.empty, logger: Option[ExecutionLogger] = None)(
    implicit sc: ExecutionStateTracker = new ExecutionStateTracker
  ): AsyncExecution[T] = {
    new Logic(dataFlows).asyncRun(token, extraFlow, logger)(sc)
  }

  def ++(another: Niffler): Niffler = new Niffler {
    override def dataFlows: Iterable[DataFlow[_]] = {
      self.dataFlows ++ another.dataFlows
    }
  }

  def printGraph(logger: Logger, useCodeName: Boolean): Unit = {
    new Logic(dataFlows).printFlowChart(logger, Level.INFO, useCodeName)
  }

  def printGraph(logger: Logger, useCodeName: Boolean, extraFlow: Iterable[DataFlow[_]]): Unit = {
    new Logic(dataFlows ++ extraFlow).printFlowChart(logger, Level.INFO, useCodeName)
  }
}
